provider/common/bootstrap.go

```diff
@@ -180,11 +181,41 @@ func BootstrapInstance(ctx environs.BootstrapContext, env environs.Environ, args
 		ImageMetadata:   imageMetadata,
 		StatusCallback:  instanceStatus,
 		CleanupCallback: statusCleanup,
-	})
-	if err != nil {
+	}
+	zones, err := startInstanceZones(env, startInstanceArgs)
+	if errors.IsNotImplemented(err) {
+		// No zone support, so just call StartInstance with
+		// a blank StartInstanceParams.AvailabilityZone.
+		zones = []string{""}
+	} else if err != nil {
 		return nil, "", nil, errors.Annotate(err, "cannot start bootstrap instance")
 	}
 
+	var result *environs.StartInstanceResult
```

Comment:
Zone selection feels like it should be part of StartInstance rather than part of Bootstrap, shouldn't it?
We need similar logic for every instance that we deploy, including load balancing, support for provider networks that only exist in particular zones, etc.

I'm a little concerned that pulling it into Bootstrap handles one case, but we have to re-implement this for all the times we go to provision an instance.

Reply:
It *was* part of StartInstance, but that doesn't work well with provisioner parallelisation. We have two places where we provision instances: bootstrap, and the provisioner. The provisioner was updated to pass in the availability zone, but the work to update bootstrap was deferred. That's this PR.